### PR TITLE
Make external update an explicit target rather then an implicit cmake step

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

**Comment by eile** [comment]
@rdumusc FYI

File: GitExternal.cmake [hunk 1]
```diff
@@ -101,67 +109,68 @@ function(GIT_EXTERNAL DIR REPO TAG)
...
+    "    ERROR_QUIET OUTPUT_QUIET\n"
+    "    WORKING_DIRECTORY \"${DIR}\")\n"
+    "  execute_process(\n"
+    "    COMMAND \"${GIT_EXECUTABLE}\" checkout -q -- \"${GIT_EXTERNAL_RESET_FILE}\"\n"
+    "    ERROR_QUIET OUTPUT_QUIET\n"
+    "    WORKING_DIRECTORY \"${DIR}\")\n"
+    "endforeach()\n"
+    "\n"
+    # fetch latest update
+    "execute_process(COMMAND \"${GIT_EXECUTABLE}\" fetch origin -q\n"
```
Comment (tribal-tec) [inline GitExternal.cmake]: make 'origin' a configurable variable? I'd like to fetch from 'user' or even '-all' sometimes.

File: GitExternal.cmake [hunk 2]
```diff
@@ -101,67 +109,68 @@ function(GIT_EXTERNAL DIR REPO TAG)
...
+    "endif()\n"
+    "\n"
+    # update tag
+    "execute_process(COMMAND ${GIT_EXECUTABLE} rebase FETCH_HEAD\n"
+    "  RESULT_VARIABLE nok ERROR_VARIABLE error OUTPUT_VARIABLE output\n"
+    "  WORKING_DIRECTORY \"${DIR}\")\n"
+    "if(nok)\n"
+    "  execute_process(COMMAND ${GIT_EXECUTABLE} rebase --abort\n"
+    "    WORKING_DIRECTORY \"${DIR}\" ERROR_QUIET OUTPUT_QUIET)\n"
+    "  message(FATAL_ERROR \"Rebase ${__dir} failed:\n\${output}\${error}\")\n"
```
Comment (tribal-tec) [inline GitExternal.cmake]: does it need to be FATAL_ERROR here?

File: GitExternal.cmake (hunk 1 above)
Comment (eile) [inline GitExternal.cmake]: New feature -> in a new commit/PR.

File: GitExternal.cmake (hunk 2 above)
Comment (eile) [inline GitExternal.cmake]: Now that it's an explicit target I'ld like it to fail on failure. One can always do 'ninja rebase -k0'

**Comment by tribal-tec** [comment]
Ok, +1 for me. I suggest you wait for @rdumusc though :)

File: GitExternal.cmake [hunk 3]
```diff
@@ -14,13 +14,12 @@
 #    or in the given absolute path using the given repository and tag
 #    (commit-ish).
 #
+# Targets:
+#  * DIR-rebase: fetches latest updates and rebases the given external
```
Comment (rdumusc) [inline GitExternal.cmake]: rename DIR-rebase to <directory>-rebase for consistency with the rest of the documentation

File: GitExternal.cmake [hunk 4]
```diff
@@ -101,67 +109,68 @@ function(GIT_EXTERNAL DIR REPO TAG)
...
+    )
 
-  # checkout requested tag
-  execute_process(
-    COMMAND "${GIT_EXECUTABLE}" checkout -q "${TAG}"
-    RESULT_VARIABLE nok ERROR_VARIABLE error
-    WORKING_DIRECTORY "${DIR}")
-  if(nok)
-    message(STATUS "git checkout ${TAG} in ${DIR} failed: ${error}\n")
+  add_custom_target(${__target}-rebase
```
Comment (rdumusc) [inline GitExternal.cmake]: Make these targets Windows friendly: EXCLUDE_FROM_DEFAULT_BUILD ON FOLDER "git" or something like that

**Comment by eile** [comment]
updated

**Comment by rdumusc** [comment]
Other than that, looks good to me! I would also propose to have all git-related target share a common prefix, such as git_rebase_all, git_update_externals, etc... but I have the feeling you won't like that idea ;-)

**Comment by eile** [comment]
definitely not with '_'. :P

Maybe in a later commit. Targets are still a mess, let's fix this one by one.

**Pull request merged**